src/compiler/checker.ts

```diff
@@ -14794,15 +14794,16 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return type.resolvedApparentType || (type.resolvedApparentType = getResolvedApparentTypeOfMappedType(type));
     }
 
-    function getResolvedApparentTypeOfMappedType(type: MappedType) {
+    function getResolvedApparentTypeOfMappedType(type: MappedType): Type {
         const target = (type.target ?? type) as MappedType;
         const typeVariable = getHomomorphicTypeVariable(target);
         if (typeVariable && !target.declaration.nameType) {
             const constraint = getConstraintTypeFromMappedType(type);
             if (constraint.flags & TypeFlags.Index) {
                 const baseConstraint = getBaseConstraintOfType((constraint as IndexType).type);
                 if (baseConstraint && everyType(baseConstraint, t => isArrayOrTupleType(t) || isArrayOrTupleOrIntersection(t))) {
-                    return instantiateType(target, prependTypeMapping(typeVariable, baseConstraint, type.mapper));

```

Comment:
The `type.mapper` here looks something like:
```ts
m1: T_id -> Box<string>[]
m2: m1: K -> K
    m2: m1: T_id -> UnboxArray<T_outer>
        m2: T_id -> UnboxArray<T_outer>
```

Since `typeVariable` is `T_id` TS ended up "skipping" over the inner mapping of `UnboxArray`. The `constraint` (of the `type`) is `keyof T_outer` with a `baseConstraint` of `Box<string>[]`.

In a sense, I'd like to map `T_outer` to `baseConstraint` (`Box<string>[]`), process that through `type.mapper` and then map `T_id` to that result (or smth along those lines). I couldn't find a clear way to do that given that we need to account for non-homomorphic instantiations of homomorphic mapped types here. Using the modifiers type does seem to do the trick but I'm not overly confident in this solution and I won't mind being proven that this is wrong 😉